nocasedict/_nocasedict.py

```diff
@@ -0,0 +1,475 @@
+                    warnings.warn(
+                        "Initializing a NocaseDict object from {0} before "
+                        "Python 3.7 will not preserve order of items".
+                        format(type(arg)),
+                        UserWarning, stacklevel=2)
+                self.update(arg)
+            elif arg is None:
+                # Leave empty
+                pass
+            else:
+                raise TypeError(
+                    "Invalid type for NocaseDict initialization: {0} ({1})".
+                    format(arg.__class__.__name__, type(arg)))
+
+        # Step 2: Add any keyword arguments
+        if kwargs:
+            if len(kwargs) > 1 and sys.version_info[0:2] < (3, 7):
+                warnings.warn(
+                    "Initializing a NocaseDict object from keyword arguments "
+                    "before Python 3.7 will not preserve order of items",
+                    UserWarning, stacklevel=2)
+            self.update(kwargs)
+
+    # Basic accessor and settor methods
+
+    def _real_key(self, key):
+        """
+        Return the normalized key to be used for the internal dictionary,
+        from the input key.
+        """
+        if key is not None:
+            try:
+                return key.lower()
+            except AttributeError:
+                raise TypeError(
+                    "NocaseDict key {0!r} must be a string, but is {1}".
+                    format(key, type(key)))
+
+        if self.allow_unnamed_keys:
+            return None
+
+        raise TypeError(
+            "NocaseDict key None (unnamed key) is not allowed for this object")
+
+    def __getitem__(self, key):
+        """
+        Invoked when retrieving the value for a key, using `val = d[key]`.
+
+        The key is looked up case-insensitively. Raises `KeyError` if the
+        specified key does not exist. Note that __setitem__() ensures that
+        only string typed keys will exist, so the key type is not tested here
+        and specifying non-string typed keys will simply lead to a KeyError.
+        """
+        k = self._real_key(key)
+        try:
+            return self._data[k][1]
+        except KeyError:
+            raise KeyError("Key {0!r} not found".format(key))
+
+    def __setitem__(self, key, value):
+        """
+        Invoked when assigning a value for a key using `d[key] = val`.
+
+        The key is looked up case-insensitively. If the key does not exist,
+        it is added with the new value. Otherwise, its value is overwritten
+        with the new value.
+
+        Raises `TypeError` if the specified key does not have string type.
+        """
+        k = self._real_key(key)
+        self._data[k] = (key, value)
+
+    def __delitem__(self, key):
+        """
+        Invoked when deleting a key/value pair using `del d[key]`.
+
+        The key is looked up case-insensitively. Raises `KeyError` if the
```

Comment:
Key lookup is case-insensitive.   

NOTE: I think this same comment applies in several more cases.

Reply:
Has already been improved in docs update PR, please review there again.